File: server/basic/guider.py
# coding=utf-8

"""

Copyright(c) 2022 Max Qian  <astroair.cn>

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public
License version 3 as published by the Free Software Foundation.
This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.
You should have received a copy of the GNU Library General Public License
along with this library; see the file COPYING.LIB.  If not, write to
the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
Boston, MA 02110-1301, USA.

"""

from server.basic.device import BasicDeviceAPI

import selectors
import socket

class TcpSocket(object):
    """
        TCP socket client interface
    """

    def __init__(self):
        self.lines = []
        self.buf = b''
        self.sock = None
        self.sel = None
        self.terminate = False

    def __del__(self):
        self.disconnect()

    def connect(self, hostname, port):
        self.sock = socket.socket()
        try:
            self.sock.connect((hostname, port))
            self.sock.setblocking(False)  # non-blocking
            self.sel = selectors.DefaultSelector()
            self.sel.register(self.sock, selectors.EVENT_READ)
        except Exception:
            self.sel = None
            self.sock = None
            raise

    def disconnect(self):
        if self.sel is not None:
            self.sel.unregister(self.sock)
            self.sel = None
        if self.sock is not None:
            self.sock.close()
            self.sock = None

    def terminate(self):
        self.terminate = True

    def read(self):
        while not self.lines:
            while True:
                if self.terminate:
                    return ''
                events = self.sel.select(0.5)
                if events:
                    break
            s = self.sock.recv(4096)
            i0 = 0
            i = i0
            while i < len(s):
                if s[i] == b'\r'[0] or s[i] == b'\n'[0]:
                    self.buf += s[i0 : i]
                    if self.buf:
                        self.lines.append(self.buf)
                        self.buf = b''
                    i += 1
                    i0 = i
                else:
                    i += 1
            self.buf += s[i0 : i]
        return self.lines.pop(0)

    def send(self, s):
        b = s.encode()
        totsent = 0
        while totsent < len(b):
            sent = self.sock.send(b[totsent:])
            if sent == 0:
                raise RuntimeError("socket connection broken")
            totsent += sent

class BasicGuiderInfo(object):
    """
        Basic Guider Information container
    """

    _is_connected = False
    _is_device_connected = False
    _is_looping = False
    _is_calibrating = False
    _is_dithering = False
    _is_guiding = False
    _is_settling = False

    _is_calibrated = False
    _is_selected = False
    _is_settled = False
    _is_starlost = False
    _is_starlocklost = False

    _can_cooling = False

    def get_dict(self) -> dict:
        """
            Return Guider Information in the dictionary format
        """

class BasicGuiderAPI(BasicDeviceAPI):
    """
        Basic Guider API
    """

    def start_guiding(self, params : dict) -> dict:
        """
            Start guiding | 开始导星
            Args:
                params:{

                }
            Returns:{
                "status" : int,
                "message" : str,
                "params" : None
            }
        """

    def abort_guiding(self) -> dict:
        """
            Abort guiding | 停止导星
            Returns:{
                "status" : int,
                "message" : str,
                "params" : None
            }
        """

    def start_calibration(self, params : dict) -> dict:
        """
            Start calibration | 开始校准
            Args:
                params:{

                }
            Returns:{
                "status" : int,
                "message" : str,
                "params" : None
            }
        """
    
    def abort_calibration(self) -> dict:
        """
            Abort the calibration | 停止校准
            Args: None
            Returns:{
                "status" : int,
                "message" : str,
                "params" : None
            }
        """

    def start_dither(self, params : dict) -> dict:
        """
            Start dithering | 开始抖动
            Args:
                params:{

                }
            Returns:{
                "status" : int,
                "message" : str,
                "params" : None
            }
        """

    def abort_dither(self) -> dict:
        """
            Abort dither | 停止抖动
            Args: None
            Returns:{
                "status" : int,
                "message" : str,
                "params" : None
            }
        """

File: client/static/js/client/guider.js
/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * 
 */

// ----------------------------------------------------------------
// Guider line Renders
// ----------------------------------------------------------------

$(function () {

    var RAData = [];
    var DECData = [];

    var GuidingLineData = {
        datasets: [
            {
                label: 'RA',
                backgroundColor: 'rgba(60,141,188,0.9)',
                borderColor: 'rgba(0,0,255,0.8)',
                pointRadius: false,
                pointColor: '#3b8bba',
                pointStrokeColor: 'rgba(60,141,188,1)',
                pointHighlightFill: '#fff',
                pointHighlightStroke: 'rgba(60,141,188,1)',
                data: RAData
            },
            {
                label: 'DEC',
                backgroundColor: 'rgba(255,0 ,0, 1)',
                borderColor: 'rgba(255, 0, 0, 0.8)',
                pointRadius: true,
                pointColor: 'rgba(210, 214, 222, 1)',
                pointStrokeColor: '#c1c7d1',
                pointHighlightFill: '#fff',
                pointHighlightStroke: 'rgba(220,220,220,1)',
                data: DECData
            },
        ]
    }

    var GuidingLineOptions = {
        maintainAspectRatio: false,
        responsive: true,
        legend: {
            display: true
        },
        scales: {
            xAxes: [{
                gridLines: {
                    display: true,
                }
            }],
            yAxes: [{
                gridLines: {
                    display: true,
                }
            }]
        }
    }


    var guiding_line_canvas = $('#guiding_line').get(0).getContext('2d')
    var guiding_line_options = $.extend(true, {}, GuidingLineOptions)
    var guiding_line_data = $.extend(true, {}, GuidingLineData)
    guiding_line_data.datasets[0].fill = false;
    guiding_line_data.datasets[1].fill = false;
    guiding_line_options.datasetFill = false

    var guiding_chart = new Chart(guiding_line_canvas, {
        type: 'line',
        data: guiding_line_data,
        options: guiding_line_options
    })
});

// ----------------------------------------------------------------
// Scatter plot Render
// ----------------------------------------------------------------

$(function () {

    var ScatterPlotOptions = {
        type: 'polarArea',
        //data: data,
        options: {
            responsive: true,
            maintainAspectRatio: false,
        }
    }

    var scatter_plot_canvas = $('#scatter_plot').get(0).getContext('2d')
    var scatter_plot_options = $.extend(true, {}, ScatterPlotOptions)
    //var scatter_plot_data = $.extend(true, {}, ScatterPlotData)

    var ScatterPlot = new Chart(scatter_plot_canvas, scatter_plot_options)
});

// ----------------------------------------------------------------
// BS-Stepper for server and device connections
// ----------------------------------------------------------------

// BS-Stepper Init
document.addEventListener('DOMContentLoaded', function () {
    window.stepper = new Stepper(document.querySelector('.bs-stepper'))
})

// ----------------------------------------------------------------
// Select2 Initialization
// ----------------------------------------------------------------

$('.select2').select2()